### PR TITLE
Added destructor function

### PR DESCRIPTION
**Pull request by XiaChuerwu**
If this PR is related to coding or code translation, please fill out the checklist and paste the console outputs to the PR.

- [x] I've tested the code and ensured the outputs are the same as the outputs of reference codes.
- [x] I've checked the codes (formatting, comments, indentation, file header, etc) carefully.
- [x] The code does not rely on a particular environment or IDE and can be executed on a standard system (Win, macOS, Ubuntu).

### CODE REVIEW

**Comment by XiaChuerwu** [comment]
image
There was a serious memory leak in the source code, which I detected using the valgrind tool under linux, and which I have now fixed by adding a destructor.
image

**Comment by XiaChuerwu** [comment]
image
image

The source code is equivalent to the original array hash table when inserting an arbitrary number of elements (twenty hellos were used for testing), which has now been fixed!

File: codes/cpp/chapter_hashing/hash_map_chaining.cpp [hunk 1]
```diff
@@ -20,10 +20,23 @@ class HashMapChaining {
     HashMapChaining() : size(0), capacity(4), loadThres(2.0 / 3), extendRatio(2) {
         buckets.resize(capacity);
     }
+    
+    /* 虚构方法 */
```
Comment (krahets) [inline codes/cpp/chapter_hashing/hash_map_chaining.cpp:24]: Is this 析构方法 ?

File: codes/cpp/chapter_hashing/hash_map_chaining.cpp [hunk 2]
```diff
@@ -20,10 +20,23 @@ class HashMapChaining {
...
+        }
+    }
 
     /* 哈希函数 */
     int hashFunc(int key) {
-        return key % capacity;
+        if(key == 0) {
+            return 0;
+        }
+        return capacity % key;
```
Comment (krahets) [inline codes/cpp/chapter_hashing/hash_map_chaining.cpp:39]: I think `capacity % key` is not correct.

File: codes/cpp/chapter_hashing/hash_map_chaining.cpp [hunk 3]
```diff
@@ -20,10 +20,23 @@ class HashMapChaining {
...
+                // 释放内存
+                delete pair;
+            }
+        }
+    }
 
     /* 哈希函数 */
     int hashFunc(int key) {
-        return key % capacity;
+        if(key == 0) {
```
Comment (krahets) [inline codes/cpp/chapter_hashing/hash_map_chaining.cpp:36]: Is this necessary?

**Review by krahets — COMMENTED** [review]
Thanks for the PR. Please address the comment.

File: codes/cpp/chapter_hashing/hash_map_chaining.cpp (hunk 1 above)
Comment (XiaChuerwu) [reply codes/cpp/chapter_hashing/hash_map_chaining.cpp:24]: Yeah, sorry about the typo.

**Comment by XiaChuerwu** [comment]
image

File: codes/c/chapter_hashing/hash_map_chaining.c [hunk 4]
```diff
@@ -0,0 +1,234 @@
+/**
```
Comment (krahets) [inline codes/c/chapter_hashing/hash_map_chaining.c:1]: Let's create a new PR for the C code addition.

**Review by krahets — APPROVED** [review]
Thanks for the fix!

1. Please create a new PR for the C code addition, which will make every PR clearer.
2. Welcome to make more fixes or code translation.

**Pull request merged**